tutorials/graph-theory/tarjans-algorithm.md

```diff
@@ -1,11 +1,346 @@
 hide_table_of_contents: true
-draft: true
 keywords:
   - leetcode
   - tutorial
   - targan's algorithm
   - algorithm
----
\ No newline at end of file
+---
+
+<TutorialAuthors names="@BlackPanther112358"/>
+
+## Overview
+
+Tarjan's Algorithm is popular algorithm for finding the Strongly Connected Components (SCC) of a directed graph. In the below graph, the nodes in a blue envelope constitute a single connected component as any node $u$ as a path to another node $v$ and vice versa. Note that while we can reach the node $f$ from the node $e$, the opposite is not true. Hence, $f$ and $e$ are not part of the same component. Thus, the following graph has 3 strongly connected components as highlighted. A single strongly connected component can be formally described as maximal set of vertices such that for any 2 vertices belonging to the set, say $u$ and $v$, there exists a path from $u$ to $v$ and vice versa.
+
+
+![image](https://user-images.githubusercontent.com/83649829/198752836-3c43a6f4-e1e2-445e-b679-0df4cb8ff3e0.png)$\\$
+[Source](https://en.wikipedia.org/wiki/Strongly_connected_component)
+
+
+In this tutorial we will discuss the Tarjan's Algorithm to find SCC.
+
+## Implementation
+
+The algorithm works by using a single [DFS](../../tutorials/graph-theory/depth-first-search) in the graph. To understand its working first let's define a few standard terms :-
+
+$tin(u) =$ time at which node $u$ was reached for the first time or the in time of DFS for the node.
+
+$low(u) =$ smallest time reachable of a node reachable from the DFS subtree of node u
+
+We also need to ensure that we don't mix 2 different SCC's due to a cross-edge between them. To counter this issue, we will use a stack to store the nodes which have not been assigned to an SCC and have been visited so far.
+
+Thus, whenever we find an SCC, we will pop the corresponding nodes from the stack.
+
+We will call the first node we discovered of an SCC the route node for sake of simplicity. Thus, we will identify the SCC by its root node. To check when we have reached a root node, we just check if $low(u)$ and $tin(u)$ are equal.
+
+The pseudo-code for the algorithm can be found [here](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm#The_algorithm_in_pseudocode)
+
+The Time and Space Complexity of the algorithm is $O(V + E)$, where $V$ represents the number of vertices and $E$ represents the number of edges, same as that os DFS.
+
+The implementation of above can be as follows (along with above graph as example) :
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+struct tarjans_algo{
+    // Recursive function to do the DFS search on the graph
+    void dfs(int u, int &timer, vector<int> &tin, vector<int> &low_link, vector<bool> &on_stack, stack<int> &stk, vector<vector<int>> &graph, vector<vector<int>> &res){
```

Comment:
this line is too long. ideally we don't let readers to scroll

Reply:
I need all of these datastructures to run the algorithm, the other option is to make these global, which I am doing for now. If you have something else in mind, do let me know

Reply:
Either breaking them into multiple lines or making some of them global

Reply:
I have made all of them except adjacency list global, do have a look once you find time